### DISCUSSION TITLE
How to constrain the overall renewable share for all nodes?

### DISCUSSION THREAD

**Discussion by MiaoCheng0517** [iiasa/message_ix#875] [Q&A] [answered]
Dear support team and everyone, I met a question which is slimilar to https://github.com/iiasa/message_ix/discussions/845

I need to add a constraint on the overall renewable share of all nodes. I use the set `map_spatial_hierarchy` mentioned by @behnam-zakeri in the link above. However, it does not work for my model.

Specifically, I assigned two spatial levels in my model, as you can see in the following code: 

```
scenario.add_set("node", "China")
scenario.add_set("node", province)
scenario.add_set("lvl_spatial", ["country", "province"])
scenario.add_set("map_spatial_hierarchy", ["country", "China", "World"])
for pro in province:
    scenario.add_set("map_spatial_hierarchy", ["province", pro, "China"])
scenario.set("map_spatial_hierarchy")
```

where `province` contains several names of China's province. For example, `province = ["Beijing","Tianjin","Shanghai","Chongqing"]`

Refering to https://github.com/iiasa/message_ix/blob/main/tutorial/westeros/westeros_share_constraint.ipynb, I could add constraints on a single node, take Chongqing for example, the code is as follows:
```
df = pd.DataFrame(
    {
        "shares": ["share_renewable_electricity"],
        "node_share": "Chongqing",
        "node": "Chongqing",
        "type_tec": "electricity_total",
        "mode": "standard",
        "commodity": "electricity",
        "level": "secondary",
    }
)
scenario.add_set("map_shares_commodity_total", df)
df = pd.DataFrame(
    {
        "shares": ["share_renewable_electricity"],
        "node_share": "Chongqing",
        "node": "Chongqing",
        "type_tec": "electricity_renewable",
        "mode": "standard",
        "commodity": "electricity",
        "level": "secondary",
    }
)
scenario.add_set("map_shares_commodity_share", df)
df = pd.DataFrame(
    {
        "shares": "share_renewable_electricity",
        "node_share": "Chongqing",
        "year_act": [2068, 2073],
        "time": "year",
        "value": [0.9, 0.95],
        "unit": "-",
    }
)
scenario.add_par("share_commodity_lo", df)
```
With these setting, I can observe that the renewable share in `"Chongqing"` is limited to 0.9 and 0.95 in 2068 and 2073, respectively.

To constrain the overall renewable share in China, I substituted every `"Chongqing"` (totally five) in the above block with `"China"`. However, the output is not as wish. The overall renewable share in all province is obviously exceed 0.9 in 2068 and 0.95 in 2073.

Thank you very much for your reading and answering!

**Comment by glatterf42** [comment]
Hi @MiaoCheng0517, thanks for this question :)
Thanks for searching the tutorials, I think the one you found is the right one for this case, if our code supports this. Please clarify for me: what kind of bound do you want to add for the renewable share?
The code you posted here is adding adding a `share_commodity_lo`, which defines a lower bound. If you want to define an upper bound instead, please try using `share_commodity_up`. If you want to limit the model to these exact numbers, you can add both -- the upper and the lower bound.

**Reply by MiaoCheng0517** [reply]
Thank you for your response. I want to define the lower bound of renewable share. But the renewable share constraint on the parent node doesn't seem to have been applied.

I believe the confusion was caused by an error in my description, which led to a misunderstanding. The overall renewable share in all procinve is obviously **lower than** 0.9 in 2068 and 0.95 in 2073. I apologize for my carelessness.

**Comment by khaeru** [comment]
@MiaoCheng0517, it is hard for us to know what precisely is happening and provide advice without knowing the exact contents/structure of the model you have built.

However, it is possible to make some guesses. My guess here is you are (reasonably) expecting a feature that the GAMS implementation of MESSAGE does not have.

In particular, you put constraints on technologies that output to (that is: have `output` parameter values for):
- commodity: electricity
- level: secondary
- node: Chongqing

These technologies are in two groups called `electricity_renewable` and `electricity_total`. (You don't say what technologies you have put in these groups.)

This activates the equation called `SHARE_CONSTRAINT_COMMODITY_LO`, which is documented [here](https://docs.messageix.org/en/latest/model/MESSAGE/model_core.html#equation-share-constraint-commodity-lo).

Then, you change to n="China". Separately, you have used `map_spatial_hierarchy` to put n="Chongqing" and some other nodes (your message says `for pro in province:` but you don't show what `province` is) under n="China".

Two key points here:
1. Again, I am guessing, but you **do not** have any technologies (in the `electricity_renewable` or `electricity_total` group) that output to:
   - commodity: electricity
   - level: secondary
   - **node: China**

   In other words, there is no data in `output` with these keys. Rather, I guess, your technologies *all* output to their respective `node_dest`, which is the same as `node_loc`, e.g. "Chongqing".

   If this is not the case, please provide further information.

2. The equation SHARE_CONSTRAINT_COMMODITY_LO **does not automatically** compute *aggregates* over the commodity outputs to "spatial hierarchy levels" that are lower than any respective level. You can confirm this by:
   - Looking at the documentation page I linked above. Notice that `n^L` appears as a subscript—i.e. `output` and other parameter/variable data are handled for particular nodes, but **no mention** is made of "child" nodes at lower.
   - Looking at the GAMS code directly: https://github.com/iiasa/message_ix/blob/00241f6fea4c689ee7f7173cd62b07d218f6571b/message_ix/model/MESSAGE/model_core.gms#L1485-L1517 See that it does not identify the sub-nodes of any given node and compute their activity/output/input.

In other words, because of (1), your model does not output anything to (n=China), and so the constraints have no effect.

Let's please confirm this is a correct understanding of your situation. After that, we can talk about how to achieve what you intend.

**Reply by glatterf42** [reply, marked as answer]
Thanks for jumping in here, @khaeru. I have to say, though: I was also just studying the docs and this exact GAMS code and would not have arrived at your conclusions. In my view, this equation is not well documented, in particular, I don't see any documentation of the required `map_shares_commodity_total` and `map_shares_commodity_share`. So I, for one, don't understand why you can specify `node_share` and `node` and what these mean in this context. From our docs, "The implementation is generic and flexible, so that any combination of commodities, levels, technologies and nodes can be put in relation to any other combination" makes it sound like our code *can* handle the desired use case here. My interpretation would have been that we just need to figure out the right combination of `node_share` and `node`.
Maybe @behnam-zakeri again has more insight into the spatial implementation?

**Reply by MiaoCheng0517** [reply]
Thanks @khaeru @glatterf42 for seriously considering the questions I raised.

1. @khaeru  's conjecture about my model was correct. My technologies all output to their respective `node_dest`, which is the same as `node_loc`, e.g. "Chongqing". No technology output to the parent node `"China"`.
2. According to @glatterf42 's insight, I changed all `node_share` in sets `"map_shares_commodity_total"` and `"map_shares_commodity_renewable"` to the parent node `"China"`, and kept all `node` in these two sets as individual child node, e.g. `"Chognqing"`. The `node_share` in the parameter `"share_commodity_lo"` was assigned as `"China"`. Then it worked. The output shows that the renewable share of each child node might not necessarily exceed 0.9, but the aggregate exceeds 0.9.

Thank you again for your response and discussion.
